File: 0242-valid-anagram/0242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:

        if len(s) != len(t):
            return False
        
        letters = {}

        for c in s:
            letters[c] = letters.get(c, 0) + 1

        for c in t:
            if c not in letters:
                return False
            letters[c] -= 1
            if letters[c] == 0:
                del letters[c]

        return True

File: 0242-valid-anagram/0242-valid-anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {

        if(s.length() != t.length())
            return false;

        Map<Character, Integer> chars = new HashMap<>();

        for(char c : s.toCharArray()){
            chars.put(c, chars.getOrDefault(c, 0) + 1);
        }

        for(char c : t.toCharArray()){
            if(!chars.containsKey(c)){
                return false;
            }
            chars.put(c, chars.get(c)-1);
            if(chars.get(c) == 0){
                chars.remove(c);
            }
        }        
        return true;
    }
}